### ISSUE TITLE
Performance/Bug: Detach PyTorch tensors in CallbackManager payloads to prevent GPU VRAM leaks

### ISSUE THREAD

**Issue by Neelr1912** [Adit-Jain-srm/NightmareNet#159]
### Describe the problem
With the recent integration of `CallbackManager` into the `Trainer` [2.1.7], the system now emits fine-grained `TrainingEvent` objects (like `STEP`, `METRIC`, and `EPOCH_END`). If raw PyTorch loss or accuracy tensors are passed directly into these event payloads (which are then consumed by `pipeline.py`, WebSockets, or webhooks), it retains the entire computation graph in memory [2.1.7]. Over a long robustness training cycle or evaluation sweep [2.1.5], this will rapidly exhaust GPU VRAM, leading to a `CUDA OutOfMemoryError`.

### Expected behavior
Any metric or state payload emitted by the `Trainer` or `Evaluator` to the `CallbackManager` must be fully decoupled from PyTorch's autograd engine and transferred to CPU memory before being broadcast to subscribers.

### Proposed Solution
1. In `nightmarenet/training/trainer.py` [2.1.7], intercept metrics before they are packaged into the `TrainingEvent` payload.
2. Create a utility function (e.g., `sanitize_metrics()`) that iterates over the payload dict and safely calls `.detach().cpu().item()` on any active `torch.Tensor`.
3. Apply this sanitization step before emitting `EPOCH_START`, `EPOCH_END`, `STEP`, and `METRIC` events.
4. Add a unit test in `tests/test_callbacks.py` [2.1.7] that asserts event payloads contain native Python floats/types rather than PyTorch tensors with `requires_grad=True`.

### Scope / Difficulty Estimate
**Difficulty:** Low/Medium
**Files affected:**
- `nightmarenet/training/trainer.py`
- `nightmarenet/evaluation/evaluator.py`
- `tests/test_callbacks.py`

**Comment by Diwakar-odds** [comment]
### Assignment Request (L3 / Core Performance Optimization)
I would like to take on this performance fix. Below is my proposed technical implementation plan to resolve the VRAM leaks.

**Step-by-Step Implementation Plan:**
1. **Payload Inspection:** Review the `CallbackManager` logic to identify where model outputs (which are attached to the computation graph) are being passed to callbacks (like logging, metric computation, or visualization).
2. **Tensor Detachment:** Introduce a utility function that recursively traverses the payload dictionary/list and calls `.detach().cpu()` on any PyTorch tensors before they are handed off to the callbacks.
3. **Memory Profiling Test:** Add a unit test that runs a mini training loop with mock callbacks, measuring VRAM usage before and after to assert that memory does not monotonically increase over epochs.

**Files to be modified:**
- `nightmarenet/callbacks/manager.py` (or equivalent callback handler)
- `nightmarenet/utils/tensor_utils.py` (New: utility for recursive detach)
- `tests/test_callbacks_memory.py`

I'll ensure all code passes `ruff` and `mypy` as required by the contribution rules. Let me know if I can start!

**Comment by panditshubham766-dotcom** [comment]
Hi @Adit-Jain-srm , I'd like to work on this issue!
My plan:
After reviewing the codebase, here's my implementation approach:

1)Add a sanitize_metrics() utility function in nightmarenet/training/trainer.py that recursively traverses the payload dict and calls .detach().cpu().item() on any active torch.Tensor before it's packaged into a TrainingEvent
2)Apply sanitize_metrics() before emitting EPOCH_START, EPOCH_END, STEP, and METRIC events in nightmarenet/training/trainer.py
3)Apply the same sanitization in nightmarenet/evaluation/evaluator.py before emitting evaluation metrics
4)Add a unit test in tests/test_callbacks.py that creates mock tensors with requires_grad=True, passes them through the callback pipeline, and asserts the received payload contains native Python floats/ints — not PyTorch tensors

Files touched: nightmarenet/training/trainer.py, nightmarenet/evaluation/evaluator.py, tests/test_callbacks.py
Could you please assign this to me? Thanks!

**Comment by Adit-Jain-srm** [comment]
After investigating the actual code paths, **this bug does not exist in the current codebase.** Here's the evidence:

**All metrics are Python floats before reaching CallbackManager:**

In `nightmarenet/training/phases.py`, every phase calls `.item()` on the loss tensor immediately:

```python
# WakePhase (line ~111):
epoch_loss += loss.item() * self.gradient_accumulation_steps

# DreamPhase (line ~282):
epoch_loss += loss.item() * self.gradient_accumulation_steps
epoch_kl += kl_loss.item()

# NightmarePhase (line ~478):
epoch_loss += loss.item() * self.gradient_accumulation_steps
```

The values emitted to `TrainingEvent.metrics` are computed from these accumulated Python floats:

```python
avg_epoch_loss = epoch_loss / max(step_count, 1)  # float / int = float

# What gets emitted:
self.callback_manager.emit(
    TrainingEvent(
        event_type=EventType.EPOCH_END,
        metrics={"avg_loss": avg_epoch_loss},  # Python float, not a tensor
    )
)
```

And the phase return dicts (consumed by `trainer.py` for PHASE_END events):

```python
return {
    "phase": "wake",
    "avg_loss": total_loss / max(num_epochs, 1),  # float / int = float
}
```

**No computation graph is retained. No VRAM leak occurs.**

**Additional inaccuracies in this issue:**
- References "[2.1.7]" - no such version exists (project is at v0.2.0)
- Claims "STEP" events are emitted - they are not. Only PHASE_START/END, EPOCH_START/END, and CHECKPOINT events exist in the current implementation
- Claims "METRIC" events are emitted - also not present

**Conclusion:** The issue describes a hypothetical vulnerability that the existing code already prevents via `.item()` calls. A preventive `sanitize_metrics()` guard could be added as a defensive measure, but this is a low-priority enhancement, not a bug or performance issue.

Closing as not-a-bug. If anyone wants to contribute a defensive `sanitize_metrics()` utility as a hardening enhancement, please open a new issue with correct code references and the appropriate framing.
